### ISSUE TITLE
JSON scalar types not supported

### ISSUE THREAD

**Issue by greemo** [wundergraph/wundergraph#448]
### Bug description

I have an upstream graphql server that uses JSON scalar types.
 
When the value returned for this scalar is a JSON object, wundergraph complains with the following error:

```
2022-12-14T13:53:48+11:00 ERROR apihandler/apihandler.go:801 ResolveGraphQLResponse {"component": "@wundergraph/node", "reqId": "44a4ae5f-ef82-a2e9-3731-7a0284accd0b", "error": "invalid value type 'object' for path /data/bar, expecting string, got: {\"type\":\"string\"}. You can fix this by configuring this field as Int/Float Scalar"}
github.com/wundergraph/wundergraph/pkg/apihandler.(*GraphQLHandler).ServeHTTP
        /home/runner/work/wundergraph/wundergraph/pkg/apihandler/apihandler.go:801
github.com/wundergraph/wundergraph/pkg/authentication.NewCSRFMw.func1.1
        /home/runner/work/wundergraph/wundergraph/pkg/authentication/authentication.go:864
net/http.HandlerFunc.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2084
github.com/wundergraph/wundergraph/pkg/authentication.NewLoadUserMw.func1.1
        /home/runner/work/wundergraph/wundergraph/pkg/authentication/authentication.go:655
net/http.HandlerFunc.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2084
github.com/rs/cors.(*Cors).Handler.func1
        /home/runner/go/pkg/mod/github.com/rs/cors@v1.7.0/cors.go:219
net/http.HandlerFunc.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2084
github.com/wundergraph/wundergraph/pkg/apihandler.(*Builder).BuildAndMountApiHandler.func4.1
        /home/runner/work/wundergraph/wundergraph/pkg/apihandler/apihandler.go:233
net/http.HandlerFunc.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2084
github.com/wundergraph/wundergraph/pkg/apihandler.logRequestMiddleware.func1.1
        /home/runner/work/wundergraph/wundergraph/pkg/apihandler/apihandler.go:353
net/http.HandlerFunc.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2084
github.com/gorilla/mux.(*Router).ServeHTTP
        /home/runner/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210
net/http.serverHandler.ServeHTTP
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:2916
net/http.(*conn).serve
        /opt/hostedtoolcache/go/1.18.8/x64/src/net/http/server.go:1966
```

### How to reproduce

I have been able to create a self-contained wundergraph project to reproduce the error, using the in-built graphql server. It is attached: [wgtest.zip](https://github.com/wundergraph/wundergraph/files/10223973/wgtest.zip)

To reproduce:

- install dependencies with `yarn`
- run with `yarn start`

Then visit the following URL: `http://localhost:9991/operations/Hello`

### Expected behavior

The response from the server should be successfully returned to the caller.

For the attached project, that is:
`
{
  "data": {"bar": {"type": "string"}}
}
`

### WunderGraph information

Included in the attached project

### Environment & setup

- OS: Mint w/kernel 5.4.0-135-generic
- Go version: go version go1.13.8 linux/amd64
- Node.js version: v18.12.1

### WunderCtl Version

119.1

**Comment by jensneuse** [comment]
I think we know the problem. Is this happening with a GraphQL datasource? or Prisma? OAS?

**Comment by Knorway** [comment]
facing the same issue here. I use graphql data sources(postgraphile and hasura). it console logs the correct data including json field

**Comment by greemo** [comment]
Yep, graphql data source..

**Comment by greemo** [comment]
I also get this error when the graphql server returns an JSON int literal for the value.

```
2022-12-15T16:57:10+11:00 ERROR apihandler/apihandler.go:2257 ResolveGraphQLResponse failed {"component": "@wundergraph/node", "operationName": "myOperation", "operationType": "QUERY", "error": "invalid value type 'number' for path /data/somewhere/value, expecting string, got: 0. You can fix this by configuring this field as Int/Float Scalar"}
```

**Comment by jensneuse** [comment]
We know the solution and are currently implementing it.

**Comment by greemo** [comment]
Thanks guys. Just as an FYI, this is also an issue for input variables on mutations. If an input variable is of type JSON scalar, wundergraph converts it to a string before passing it to the graphql datasource.

**Comment by jensneuse** [comment]
@spetrunin this makes sense, we should tackle it too!

**Comment by spetrunin** [comment]
@greemo could you also provide us with  an example setup for mutations?

**Comment by greemo** [comment]
Hi @spetrunin I am having trouble setting up a mutation resolver that accepts a JSON argument. In theory, I think the below should point out the issue, but I'm getting a problem from wundergraph when trying to run it:

`"Couldn't configure your WunderNode: Error: error normalizing operation: external: Unknown type \"\"., locations: [], path: [mutation], operationFilePath: /.../wgtest/.wundergraph/operations/write.graphql"}`

If you can see the issue, great, otherwise, I'll try setting up a demo server in a different way to illustrate the problem. 

Or do you just need an example SDL file and an example operation to illustrate the scenario I am referring to?

Here is the project with the mutation operation and resolver I "think" should work, but which I am having troubles with.
 
[wgtest.zip](https://github.com/wundergraph/wundergraph/files/10314072/wgtest.zip)

**Comment by spetrunin** [comment]
@greemo in your zip file you have defined mutation like this
```graphql
mutation writeit($input: String) {
    writeit(input: $input)
}
```
but variable should be with a json type
```graphql
mutation writeit($input: JSON) {
    writeit(input: $input)
}
```

**Comment by greemo** [comment]
Thanks @spetrunin. I thought I had tried that. Oh well, it works now...

I made a small change so the operation returns the type of the argument received.

`curl -X POST -H "Content-Type: application/json" -d '{"input": 23}' http://localhost:9991/api/main/operations/write`

Returns: `{"data":{"writeit":"string"}}`

Here is the new project: [wgtest.zip](https://github.com/wundergraph/wundergraph/files/10323985/wgtest.zip)

Thanks for your help

**Comment by auth00** [comment]
I am getting a similar error as this issue describes using an oas backend. Is the solution that is mentioned in this issue included in in the 0.130.2 release of @wundergraph/sdk?

I have added a generated openapi spec from https://www.drupal.org/project/openapi_jsonapi as a backend to wundergraph and any request to an entity in Drupal returns the error: (the response is shortened with `...` for readability) 

`invalid value type 'object' for path /data/drupalapi_getNodePage, expecting string, got: {\"jsonapi\":{\"version\":\"1.0\",\"meta\":{\"links\":{\"self\":{\"href\":\"http:\\/\\/jsonapi.org\\/format\\/1.0\\/\"}}}},\"data\":[...}}}. You can fix this by configuring this field as Int/Float Scalar"`

The identified return value of `getNodePage` is printed as `JSON` on `http://localhost:9991/graphql`

**Comment by spetrunin** [comment]
Hi @auth00 

this fix is not released yet

could you attach `wundergraph.schema.graphql` from generated folder and your open api file?

**Comment by auth00** [comment]
Thank you for clarifying the status. Please see the attached drupal-openapi-files.zip.

[drupal-openapi-files.zip](https://github.com/wundergraph/wundergraph/files/10352861/drupal-openapi-files.zip)

**Comment by spetrunin** [comment]
Thank you

We will check does the fix works with open api

**Comment by auth00** [comment]
Great! Please let me know of I can contribute with anything more.

**Comment by greemo** [comment]
Hi @spetrunin, I am partially blocked on this fix, and am wondering if I should implement a work around until this is fixed, or be patient. Is it likely this fix will be added to a release in the next few days? No pressure, I just would just like to plan from my side.

**Comment by spetrunin** [comment]
hi @greemo 
pr is under review, will be shipped soon

**Comment by greemo** [comment]
Awesome, thanks Sergiy!

**Comment by spetrunin** [comment]
@greemo `@wundergraph/sdk@0.131.0` was released with this fix

**Comment by greemo** [comment]
Hi @spetrunin There seems to be an issue with JSON strings in the latest release.

Any string such as "hello world" is a valid JSON document, as is any number, boolean or json object.

When an upstream graphql API returns a string value such as "hello world" in a response field of scalar type JSON, the outer quotes are stripped, breaking the response.

I have been able to create a self-contained wundergraph project to reproduce the error, using the in-built graphql server. The new version testing strings is attached: 
[wgtest.zip](https://github.com/wundergraph/wundergraph/files/10463518/wgtest.zip)


### To reproduce:

    install dependencies with yarn
    run with yarn start

Then visit the following URL: http://localhost:9991/operations/Hello

### Expected behavior

The response from the server should be successfully returned to the caller.

For the attached project, that is:
{ "data": {"bar": "hello world"} }

**Comment by greemo** [comment]
Thanks @spetrunin for re-opening this. Let me know if I can be of any more assistance. Cheers!

**Comment by spetrunin** [comment]
@greemo for the well-known `JSON` scalar unescape is enabled by default, so you could not turn it off
to enable it for other scalars `CustomJSON` you could provide a configuration option `customJSONScalars: ["CustomJSON"]` in the graphql introspection, so maybe it could be a solution for you to use another scalar for the fields where you don't want an unescape enabled

For now, we don't have enough feedback on this functionality

**Comment by YuriBuerov** [comment]
@greemo Thanks for the bug report. The fix was released: `@wundergraph/sdk@0.133.0`. We would greatly appreciate any feedback you have.
